### DISCUSSION TITLE
Need to deploy Trivy operator in OCI  ( OKE ) ,

### DISCUSSION THREAD

**Discussion by santoshkarp3** [aquasecurity/trivy-operator#2079] [Help & Support]
Need to deploy Trivy operator in OCI  ( OKE ) , 
1. its air gapped environment  ( created internal repo for triviydb  ) 
2. with trivy we can able to scan the image 
3. but when I use with helm its not working ( pod/scan-vulnerabilityreport pod not turn into running state  )
4. Error -  failed to load policies 

Does any one have deployed , if yes pls share the  staeps 

or anyone can help with this error

**Comment by santoshkarp3** [comment]
Adding error ,
Only vulnerabilityreports.aquasecurity.github.io report generated but nt other report 


error for policy 
{"level":"error","ts":"2024-05-20T05:59:44Z","logger":"policyLoader.Get misconfig bundle policies","msg":"failed to load policies","error":"failed to donwload policies: failed to download built-in policies: download error: OCI repository error: 1 error occurred:\n\t* Get \"https://ghcr.io/v2/\": dial tcp 140.82.121.33:443: connect: connection timed out\n\n","stacktrace":"github.com/aquasecurity/trivy-operator/pkg/policy.(*policyLoader).GetPolicies\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/policy/loader.go:59\ngithub.com/aquasecurity/trivy-operator/pkg/policy.(*Policies).loadPolicies\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/policy/policy.go:144\ngithub.com/aquasecurity/trivy-operator/pkg/policy.(*Policies).Hash\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/policy/policy.go:114\ngithub.com/aquasecurity/trivy-operator/pkg/configauditreport/controller.(*PolicyConfigController).SetupWithManager.(*PolicyConfigController).reconcileConfig.func5\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/configauditreport/controller/policyconfig.go:129\nsigs.k8s.io/controller-runtime/pkg/reconcile.Func.Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/reconcile/reconcile.go:113\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:119\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:316\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:227"}

**Comment by chen-keinan** [comment]
@santoshkarp3 is it air gapped env ?

**Reply by santoshkarp3** [reply]
@chen-keinan  yes

**Reply by chen-keinan** [reply]
I have added support in latest trivy-operator to fallback to embedded policies see this [config](https://github.com/aquasecurity/trivy-operator/blob/main/deploy/helm/values.yaml#L544) or you can configure to download the policies package from [private registry](https://github.com/aquasecurity/trivy-operator/blob/main/deploy/helm/values.yaml#L665)

**Comment by santoshkarp3** [comment]
@chen-keinan 
Thanks a lot for reply ..

I am testing the above option 

One querry ..
I have put the below 2 files in private db repo , do I have to add any other file ?

  - ghcr.io/aquasecurity/trivy-db:2
  - ghcr.io/aquasecurity/trivy-java-db:1

**Reply by chen-keinan** [reply]
yes you'll need also to add [policies bundle](https://github.com/aquasecurity/trivy-operator/blob/main/deploy/helm/values.yaml#L665)

**Reply by santoshkarp3** [reply]
@chen-keinan  

just like this  ?
  registry: ghcr.io/aquasecurity/trivy-checks:0

**Reply by chen-keinan** [reply]
yes

**Reply by santoshkarp3** [reply]
@chen-keinan 
 ghcr.io/aquasecurity/trivy-checks:0 is image  or database? 
is it dynamic or changing every 6 hours  ?

**Reply by chen-keinan** [reply]
its rego policies I would say it change no earlier then once a week

**Reply by santoshkarp3** [reply]
thanks @chen-keinan 

so it will be downloaded by  oras  same as other DB files

**Reply by chen-keinan** [reply]
Same as other

**Reply by santoshkarp3** [reply]
@chen-keinan 
Sorry for late reply.

I am still getting error 

{"level":"error","ts":"2024-05-30T05:00:17Z","logger":"policyLoader.Get misconfig bundle policies","msg":"failed to load policies","error":"failed to download policies: failed to download built-in policies: download error: repository name error (vf-gks-trivy-artifacts/aquasecurity/trivy-checks:0/aquasecurity/trivy-checks:0): could not parse reference: vf-gks-trivy-artifacts/aquasecurity/trivy-checks:0/aquasecurity/trivy-checks:0","stacktrace":"github.com/aquasecurity/trivy-operator/pkg/policy.(*policyLoader).GetPolicies\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/policy/loader.go:61\ngithub.com/aquasecurity/trivy-operator/pkg/policy.(*Policies).loadPolicies\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/policy/policy.go:144\ngithub.com/aquasecurity/trivy-operator/pkg/policy.(*Policies).Eval\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/policy/policy.go:199\ngithub.com/aquasecurity/trivy-operator/pkg/configauditreport/controller.evaluate\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/configauditreport/controller/helper.go:45\ngithub.com/aquasecurity/trivy-operator/pkg/configauditreport/controller.(*ResourceController).SetupWithManager.(*ResourceController).reconcileResource.func11\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/configauditreport/controller/resource.go:229\nsigs.k8s.io/controller-runtime/pkg/reconcile.Func.Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.2/pkg/reconcile/reconcile.go:113\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.2/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.2/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.2/pkg/internal/controller/controller.go:261\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.2/pkg/internal/controller/controller.go:222"}


Command used 

helm install trivy-operator ./deploy/helm   --namespace trivy-system   --create-namespace   --set="trivy.ignoreUnfixed=true"   --set="serviceMonitor.enabled=true"   --set="image.registry=registry-rat\.prd\.harbor\.xxxxxxx\.com/vf-gks-tkg-not-retained"   --set="trivy.image.registry=registry-rat\.prd\.harbor\.xxxxxxx\.com/vf-gks-tkg-not-retained"   --set="nodeCollector.registry=registry-rat\.prd\.harbor\.xxxxxxx\.com/vf-gks-tkg-not-retained"   --set="trivy.nonSslRegistries.pocRegistry=registry-rat.prd.harbor.xxxxxxx.com"   --set="trivy.nonSslRegistries.internalRegistry=registy-rat.prd.harbor.xxxxxxx.com"   --set="trivy.insecureRegistries.pocRegistry=registry-rat.prd.harbor.xxxxxxx.com"   --set="trivy.insecureRegistries.internalRegistry=registy-rat.prd.harbor.xxxxxxx.com"   --set="trivyOperator.scanJobCompressLogs=false"   --set="trivy.dbRepositoryInsecure=true"   --set="trivy.dbRegistry=registry-rat.prd.harbor.xxxxxxx.com"   --set="trivy.dbRepository=vf-gks-trivy-artifacts/aquasecurity/trivy-db:2"   --set="trivy.javaDbdbRegistry=registry-rat.prd.harbor.xxxxxxx.com"   --set="trivy.javaDbdbRepository=vf-gks-trivy-artifacts/aquasecurity/trivy-java-db:1"  --set="policiesBundle.registry=registry-rat.prd.harbor.xxxxxxx.com" --set="policiesBundle.registry=vf-gks-trivy-artifacts/aquasecurity/trivy-checks:0"   --set="trivy.useBuiltinRegoPolicies=false"   --set="trivy.useEmbeddedRegoPolicies=true"


this is my trivy-db repo 
vf-gks-trivy-artifacts/aquasecurity/trivy-checks
vf-gks-trivy-artifacts/aquasecurity/trivy-db
vf-gks-trivy-artifacts/aquasecurity/trivy-java-db


Could you please suggest ?

**Reply by chen-keinan** [reply]
@santoshkarp3 at the latest `trivy-operator` release v0.21.1 has have added [embedded-policies](https://github.com/aquasecurity/trivy-operator/blob/main/deploy/helm/values.yaml#L544) support.

please try it up and let mw know if it solve you issue

**Reply by santoshkarp3** [reply]
I am already using 

version: 0.23.1

appVersion: 0.21.1

**Reply by chen-keinan** [reply]
so you''ll need to set up the flag I mention above

**Reply by santoshkarp3** [reply]
@chen-keinan , thanks for reply .
I used below flag already ( command is in above reply ) 
--set="trivy.useBuiltinRegoPolicies=false" --set="trivy.useEmbeddedRegoPolicies=true"

Do I miss anything?

**Reply by chen-keinan** [reply]
strange if `useBuiltinRegoPolicies == false` it should not try to download policies as describe in [code](https://github.com/aquasecurity/trivy-operator/blob/main/pkg/policy/policy.go#L143)
can you make sure that `trivy-operator-trivy-config` has this value set up

**Reply by santoshkarp3** [reply]
@chen-keinan  

It seems config map is there and below are its values 

[opc@sn-c1-10-114-14-0-26 trivy-operator-main]$ k -n trivy-system get cm  trivy-operator-trivy-config -o yaml 
apiVersion: v1
data:
  trivy.additionalVulnerabilityReportFields: ""
  trivy.command: image
  trivy.dbRepository: registry-rat.prd.harbor.xxxxxxxx.com/vf-gks-trivy-artifacts/aquasecurity/trivy-db:2
  trivy.dbRepositoryInsecure: "true"
  trivy.filesystemScanCacheDir: /var/trivyoperator/trivy-db
  trivy.ignoreUnfixed: "true"
  trivy.imagePullPolicy: IfNotPresent
  trivy.imageScanCacheDir: /tmp/trivy/.cache
  trivy.includeDevDeps: "false"
  trivy.insecureRegistry.internalRegistry: registy-rat.prd.harbor.xxxxxxxx.com
  trivy.insecureRegistry.pocRegistry: registry-rat.prd.harbor.xxxxxxxx.com
  trivy.javaDbRepository: ghcr.io/aquasecurity/trivy-java-db
  trivy.mode: Standalone
  trivy.nonSslRegistry.internalRegistry: registy-rat.prd.harbor.xxxxxxxx.com
  trivy.nonSslRegistry.pocRegistry: registry-rat.prd.harbor.xxxxxxxx.com
  trivy.offlineScan: "true"
  trivy.repository: registry-rat.prd.harbor.xxxxxxxx.com/vf-gks-tkg-not-retained/aquasecurity/trivy
  trivy.resources.limits.cpu: 500m
  trivy.resources.limits.memory: 500M
  trivy.resources.requests.cpu: 100m
  trivy.resources.requests.memory: 100M
  trivy.sbomSources: ""
  trivy.severity: UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL
  trivy.skipJavaDBUpdate: "false"
  trivy.slow: "true"
  trivy.supportedConfigAuditKinds: Workload,Service,Role,ClusterRole,NetworkPolicy,Ingress,LimitRange,ResourceQuota
  trivy.tag: 0.51.2
  trivy.timeout: 5m0s
  trivy.useBuiltinRegoPolicies: "true"
kind: ConfigMap
metadata:
  annotations:
    meta.helm.sh/release-name: trivy-operator
    meta.helm.sh/release-namespace: trivy-system
  creationTimestamp: "2024-05-30T07:21:38Z"
  labels:
    app.kubernetes.io/instance: trivy-operator
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: trivy-operator
    app.kubernetes.io/version: 0.21.1
    helm.sh/chart: trivy-operator-0.23.1
  name: trivy-operator-trivy-config
  namespace: trivy-system
  resourceVersion: "4424420"
  uid: 6e77363d-fb77-41f2-91db-6526c77d4c16
[opc@sn-c1-10-114-14-0-26 trivy-operator-main]$

**Reply by chen-keinan** [reply]
@santoshkarp3 I see in your configmap that `trivy.useBuiltinRegoPolicies: "true"` where is should be `false` and `useEmbeddedRegoPolicies` is not present

**Reply by santoshkarp3** [reply]
@chen-keinan 
I have edited cm manually and   I can see few reports generated via CLI.

BIG Thank you @chen-keinan

**Reply by santoshkarp3** [reply]
@chen-keinan  
clusterconfigauditreports and  infraassessmentreports are not genrated ,  can you please advice ?

**Reply by santoshkarp3** [reply]
hello @chen-keinan  
I am passing this with helm command, but its getting ignored 

  --set="trivy.useBuiltinRegoPolicies=false" \
  --set="trivy.useEmbeddedRegoPolicies=true" \

**Reply by chen-keinan** [reply]
I'll check and update you

**Reply by chen-keinan** [reply]
@santoshkarp3 thanks for the feedback, I see where the issue now.
converting this discussion to issue #2113
